### DISCUSSION TITLE
Javascript Client for Frontend

### DISCUSSION THREAD

**Discussion by Koliham** [permitio/fastapi_websocket_pubsub#42] [Ideas]
I was able to extract the payloads sent between client and server to create a Javascript client with websocket (code below).

But there are 2 issues, which I haven't solved yet:
1. While working fine on Windows and Android Browsers, there are issues on Safari: The connection to the Websocket is established, but the browser doesn't receive the messages, when triggering pub.
2. The connection pool doesn't seem to be cleaned up, when a client disconnects. I get these error messages, because a subscriber was not found:

<details>
  <summary><h3>Click to see stack trace!</h3></summary>
  
  ```
Failed to notify subscriber sub_id=3d0477159cdc4eb0bd8afefa9df82423 with topic=topicA
Traceback (most recent call last):
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/fastapi_websocket_pubsub/event_notifier.py", line 220, in callback_subscribers
    await self.trigger_callback(data, topic, subscriber_id, event)
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/fastapi_websocket_pubsub/event_notifier.py", line 178, in trigger_callback
    await subscription.callback(subscription, data)
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/fastapi_websocket_pubsub/event_broadcaster.py", line 159, in __broadcast_notifications__
    async with self._sharing_broadcast_channel:
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/broadcaster/_base.py", line 48, in __aenter__
    await self.connect()
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/broadcaster/_base.py", line 55, in connect
    await self._backend.connect()
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/broadcaster/_backends/postgres.py", line 13, in connect
    self._conn = await asyncpg.connect(self._url)
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/asyncpg/connection.py", line 2093, in connect
    return await connect_utils._connect(
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 903, in _connect
    raise last_error
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 889, in _connect
    return await _connect_addr(
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 781, in _connect_addr
    return await __connect_addr(params, timeout, True, *args)
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 833, in __connect_addr
    tr, pr = await compat.wait_for(connector, timeout=timeout)
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/asyncpg/compat.py", line 66, in wait_for
    return await asyncio.wait_for(fut, timeout)
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/asyncio/tasks.py", line 481, in wait_for
    return fut.result()
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 700, in _create_ssl_connection
    do_ssl_upgrade = await pr.on_data
ConnectionError: unexpected connection_lost() call
Failed to notify subscriber sub_id=3d0477159cdc4eb0bd8afefa9df82423 with topic=topicC
Traceback (most recent call last):
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/fastapi_websocket_pubsub/event_notifier.py", line 220, in callback_subscribers
    await self.trigger_callback(data, topic, subscriber_id, event)
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/fastapi_websocket_pubsub/event_notifier.py", line 178, in trigger_callback
    await subscription.callback(subscription, data)
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/fastapi_websocket_pubsub/event_broadcaster.py", line 159, in __broadcast_notifications__
    async with self._sharing_broadcast_channel:
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/broadcaster/_base.py", line 48, in __aenter__
    await self.connect()
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/broadcaster/_base.py", line 55, in connect
    await self._backend.connect()
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/broadcaster/_backends/postgres.py", line 13, in connect
    self._conn = await asyncpg.connect(self._url)
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/asyncpg/connection.py", line 2093, in connect
    return await connect_utils._connect(
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 903, in _connect
    raise last_error
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 889, in _connect
    return await _connect_addr(
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 781, in _connect_addr
    return await __connect_addr(params, timeout, True, *args)
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 833, in __connect_addr
    tr, pr = await compat.wait_for(connector, timeout=timeout)
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/asyncpg/compat.py", line 66, in wait_for
    return await asyncio.wait_for(fut, timeout)
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/asyncio/tasks.py", line 481, in wait_for
    return fut.result()
  File "/home/koliham/.cache/pypoetry/virtualenvs/websocketsample-dFaV7d37-py3.9/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 700, in _create_ssl_connection
    do_ssl_upgrade = await pr.on_data
ConnectionError: unexpected connection_lost() call
  ```
</details>

<details>
  <summary><h3>Frontend Code</h3></summary>
  
  ```vue
<template>
  <div class="text-center self-center">
       <div>Status is: {{status}} <q-btn color="primary" @click="checkStatus()">Check Status</q-btn></div>
      <div v-for="response in responses" :key="response">{{ response }}</div>
    </div>
</template>

<script>
export default {
  name: 'PageIndex',

  data() {
    return {
      inhalt: "",
      responses: [],
      status: -1
    }
  },
  created() {
    var payload = {request: {method: "subscribe", "arguments": {"topics":["topic1", "topic2", "topic3"]}}}
    var payload_str = JSON.stringify(payload)
    this.connection = new WebSocket("ws://localhost:8000/pubsub")

    this.connection.onmessage = (event) => {
      console.log(event)
      this.messageReceived(event.data)
    }

    this.connection.onopen = (event) => {
      this.connection.send(payload_str)
    }

  },

  methods: {
    checkStatus() {
      this.status = this.connection.readyState
    },
    messageReceived (content) {
    //called, when message is received
      this.responses.push(content)
    },
  }
}
</script>

  ```
</details>

<details>
  <summary><h3>Backend Code</h3></summary>
  
  ```python
import asyncio

import uvicorn
from fastapi import FastAPI
from fastapi.routing import APIRouter
from fastapi_websocket_pubsub import PubSubEndpoint
from starlette.websockets import WebSocket

PORT = 8000


app = FastAPI()
router = APIRouter()
endpoint = PubSubEndpoint(broadcaster="postgres://postgres:pw@localhost:5432/")

@router.websocket("/pubsub")
async def websocket_rpc_endpoint(websocket: WebSocket):
    async with endpoint.broadcaster:
        await endpoint.main_loop(websocket)

app.include_router(router)


async def events(topic: str = "NoTopic"):
    await endpoint.publish([topic], {"somekey": "SomeValue"})


@app.get("/trigger")
async def trigger_events():
    asyncio.create_task(events("topic1"))
    asyncio.create_task(events("topic3"))


if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=PORT)
  ```
</details>

**Comment by orweis** [comment]
Hi @Koliham :)
Kudos, on working on a lightweight JS version :)

That being said, while you got the basic schema correct; you have left out the rest of the RPC protocol, which would be my prime suspect for the issues you're experiencing.  Primarily not receiving the [responses](https://github.com/permitio/fastapi_websocket_rpc/blob/88aa59c8534b8182c85fa2a9fb839ec1d3c62bc3/fastapi_websocket_rpc/rpc_channel.py#L313) sent back for your RPC (subscribe) call

**Comment by orweis** [comment]
I've added more tips on how to implement the RPC protocol here for JS:
https://github.com/permitio/fastapi_websocket_pubsub/discussions/47#discussioncomment-3477621

**Comment by lohanspies** [comment]
Hi @orweis and @Koliham 
We managed to get a JS implementation working with `fastapi_websocket_pubsub`

Here is some example code. 
Ensure you have the `notify` event sent back to the server, otherwise, things won't work as expected.

```javascript
const WebSocket = require("ws");

try {
    socketClient = new WebSocket('ws://localhost:3010/pubsub');

    socketClient?.on("open", async () => {
        await socketClient?.send(
            '{"request": {"method": "subscribe", "arguments": {"topics": ["a", "b", "c", "d", "e", "f", "g"]}}}');

        socketClient?.on("message", async (data) => {
            try {
                const request = JSON.parse(data);
                const body = JSON.parse(request?.request?.arguments?.data);
                const call_id = request?.request?.call_id;
                console.log(`request: ${JSON.stringify(request)}`);
                console.log(`body: ${JSON.stringify(body)}`);
                if (body) {
                    console.log(JSON.stringify(body));
                    const topic = body?.topic?.toString() ?? undefined;

                    console.log(`Webhook Challenged Date: ${JSON.stringify(new Date().toISOString())} | topic: ${topic} | Body: ${JSON.stringify(body)}`);

                    try {
                        
                        const notify = {
                            "response": {
                                "result": "None",
                                "result_type": "None",
                                "call_id": call_id
                            }
                        };

                        console.log(`NOTIFYING ${JSON.stringify(notify)}`);
                        socketClient?.send(JSON.stringify(notify));

                    } catch (ex) {
                        console.log(JSON.stringify(ex));
                    }
                    //send heartbeat
                    //this.socketClient.send("heartbeat");
                    // }
                }
            } catch (ex) {
                console.log(JSON.stringify(ex));
            }
        });
    });

    socketClient?.on("error", (data) => {
        console.log(data.toString());
    });

    socketClient?.on("close", (data) => {
        console.log(data.toString());
    });
} catch (ex) {
    console.log(JSON.stringify(ex));
}
```

**Reply by orweis** [reply]
Wonderful contribution @lohanspies .

Thank you :)

**Reply by bishoyroufael** [reply]
@lohanspies  Is the notify message also needed for publishing?

**Reply by lohanspies** [reply]
Hi @bishoyroufael I am not sure but would expect that it is needed to establish the RPC connection.
Maybe @orweis can advise.

**Reply by orweis** [reply]
The "notify" message or more correctly the RPC response, is required for every RPC call - so the caller can know it was received, and stop waiting for a response.

**Comment by jhk-mjolner** [comment]
Hi, 

I'm writing my own frontend client for fastapi_pubsub and fastapi_websocket_rpc in typescript. See the files in this folder: https://github.com/Open-Pectus/Open-Pectus/tree/e68ca4e9bbc0a08563a355365c7da266cb02f069/openpectus/frontend/src/fastapi_websocket
See the test-websocket.service.ts file for usage examples.
I have it working in my test scenarios, but haven't been using it for features yet. So it's still a work in progress and things will change in the future, but given it's the summer holiday for many, it might take a while before we start using it properly ourselves.

I'm writing it as a consultant on an open source project, and we're interested in giving back to the community, but I have no plan to support it once I'm no longer attached to the project.
I _could_ pull it out into its own repo and publish an npm package, but given the above, I think it might be better in the long run if it could be a starting point for an official ts/js client npm package supported by the fastapi team.
I'm writing here, now, to start the above conversation, and to point anyone interested to my code in case they can use it.

I've written a client for RPC, one for pubsub exposing promises, and one for pubsub exposing it as RxJs Observables (useful in Angular in particular).
I had to add an 'unsubscribe' method on the python backend using the method_class= parameter to support unsubscribing in the client. See these lines: https://github.com/Open-Pectus/Open-Pectus/blob/e68ca4e9bbc0a08563a355365c7da266cb02f069/openpectus/aggregator/frontend_ws/frontend_pubsub.py#L34C1-L41C71

There's a number of features not implemented, that I might implement later, but I'm not sure I will need them all:
- No keepalive
	- I haven't seen a connection die yet, but perhaps my simple tests have been too short-lived.
- No on_connect or on_disconnect handlers 
	- I haven't needed it yet, not sure it's relevant for the js client.
- No retry logic (except for the initial ping attempt)
	- Will implement if we have the use case.
- No channel id sync
	- What is the point of this?
- No ALL_TOPICS
	- Will implement if we need it.

**Reply by orweis** [reply]
Hi @jhk-mjolner - appreciate the interest and willingness to contribute.
In a quick glance your code seems good, will need to do a more thorough CR.

I can see this being added to the projects, under another folder in the root.

- i.e. `js_websocket_pubsub` under this repo and `js_websocket_rpc` under the RPC repo
- but it should be definitely split between the two (as RPC can be also used independently)

Wouldn't mind publishing the NMP package, that's no sweat, and gradually maintaining it (Which at least initially would be mostly reviewing PRs, and publishing - as we are not focused much on this component at the time)

Regarding your unimplemented features: 

> - No keepalive - I haven't seen a connection die yet, but perhaps my simple tests have been too short-lived.

Of course, it's just a matter of time ... in long-lived connections network hiccups end up causing disconnects.
Might not be an issue if users know to handle for the case  

> - No on_connect or on_disconnect handlers - I haven't needed it yet, not sure it's relevant for the js client.

Can be useful for setting up initial subscription - which I do see you struggled with - hence the need for the added unsubscribe 
 server method.

> - No retry logic (except for the initial ping attempt) - Will implement if we have the use case.

This just gives reliability for the stack - maybe use something like https://www.npmjs.com/package/@lifeomic/attempt ?

> - No channel id sync - What is the point of this?

the channel id is used to identify clients and ongoing messages - including avoiding sending messages back to the same client when broadcasting for all.

> - No ALL_TOPICS
   - Will implement if we need it.
 
It is  useful for broadcasting a message to everyone



Let me know if you're interested in taking this the last mile

**Reply by jhk-mjolner** [reply]
Sweet!

The current plan is that I'll likely remain on the project for at least the year out, and I'll keep developing the features we need from this pubsub client in that time. So handing it over to you guys might be something we do in the autumn/winter when we're more likely to have reached a stable set of features. Unless you want it done earlier.
I'm writing this at an early stage to get the ball rolling, and maybe get some feedback.

Regarding your comments about `on_connect` and `on_disconnect`: I'm not sure what struggle you're referring to?
I have a real need for unsubscribing from the server, as the js client is a web frontend application the user will be navigating through, and topics will become relevant and irrelevant as they do. We foresee sending many notifications per second for some of our use cases, and we don't want the server spamming the client with irrelevant messages. And I don't want to disconnect the entire thing to clear up subscriptions.
The `waitForReady()` method is an adaption of your `wait_on_rpc_ready()` method, which I figured was necessary or good to have since you have it. And I suppose it makes sense to ensure there's a working rpc server on the other end of the websocket connection before making calls.
When I say I don't support `on_connect` and `on_disconnect`, I just mean I don't expose those as callbacks/events from my `RpcWebsocketClient` or pubsub clients. I _could_, I just haven't had the need yet.

About channel id sync: I'm explicitly talking about supporting that the server can get the channel id generated on the client side. From looking at the code (and I might of course have missed something) it seems like the local server-side channel id `channel.id` can be used (and _is_ used by default) in every case the client-side channel `channel.get_other_channel_id()` is used. And it's default behaviour not to sync fetch the client channel id.
I just don't understand what the use case is for getting the channel id from the client, on the server. Is it about recognizing the same channel after a disconnect-reconnect or something?

I'm interested in getting this into a state where you want to take over and support it. But I'm hoping it'll get much of the way there simply by us using it ourselves and developing the features as we need them during the coming months. And then maybe we'll add the extra little bits missing that we ourselves didn't need, but others might.

**Reply by orweis** [reply]
Hi sorry for the late reply - been a crazy week at Permit.io ;-) 

Re: connect / disconnect
The server should unsubscribe you automatically when the socket is closed - if you need to do so explicitly - that indicates some issues with the disconnection flow. -- but maybe I misread your meaning here.

- Re: unsubscribe
Aside from that it's completely fine to add methods as you see fit, that's why there's polymorphism for the RPC methods in the first place., and why the `PubSubEndpoint` receives  the `methods_class` - so one can change /add methods without having to change the internals.
That being said, And I do agree that `unsubscribe` can be useful in general - feel free to open a PR to add it the 

- > topics will become relevant and irrelevant as they do
--
we don't want the server spamming the client with irrelevant messages

  Just FYI - You should be able to design your selection of topics, to avoid such situations ; it's a matter of granularity.

- Re: waitForReady
Sounds good

- Re: channel id
  > I just don't understand what the use case is for getting the channel id from the client, on the server. Is it about recognizing the same channel after a disconnect-reconnect or something?
 
   The channel id is randomized by default, but it can also be managed by the application layer - which allows you to retain identity across connections, but keep them in account (for example for subscriptions).
   It isn't used in this code; but it could be. In addition that simple RPC flow is used to verify that the RPC connection is alive and well - which is worth the effort even just for that.
 
 
-  >I'm interested in getting this into a state where you want to take over and support it. But I'm hoping it'll get much of the way there simply by us using it ourselves and developing the features as we need them during the coming months. And then maybe we'll add the extra little bits missing that we ourselves didn't need, but others might.
 
   Sounds good, as I wrote before - aside from splitting between RPC and pub/sub code there's not a lot for you to do to get this rolling.

**Reply by jhk-mjolner** [reply]
Hi,

> sorry for the late reply

That's absolutely fine, I expect this to be a slow back-and-forth as we're probably both busy.


> The server should unsubscribe you automatically when the socket is closed - if you need to do so explicitly - that indicates some issues with the disconnection flow. -- but maybe I misread your meaning here.  

Yes, I believe there's been some minor misunderstanding, as I've had no problem with unsubscription on websocket closing. My concern about unsubscribing is all about unsubscribing certain topics (and subscribing to others) as the user navigates the frontend application.


> the PubSubEndpoint receives the methods_class - so one can change /add methods without having to change the internals.

And I'm glad it does. I was scared for a minute I had to add code to the pip-distributed files.


> That being said, And I do agree that unsubscribe can be useful in general - feel free to open a PR to add it

I expect I will do that. At least when we get into moving the JS code into the repo as well.




<blockquote>

> topics will become relevant and irrelevant as they do
    --
    we don't want the server spamming the client with irrelevant messages  

Just FYI - You should be able to design your selection of topics, to avoid such situations ; it's a matter of granularity.
</blockquote>

I'm not sure I can for my specific use cases.  
You see, I'm intending to use resource ids in the topics, to constrain the messages to be relevant to the resource the user is currently navigated to.  
To be more clear, let's say we have a resource with id 1 and another with id 2. They are both producing update messages multiple times per second. I _could_ have the frontend subscribe to a topic a la 'resource' and have it filter the messages it reacts to, to include only the resource the user is currently looking at. But that would mean a lot of unneeded and ultimately filtered-away message are sent to the frontend. And this would only increase as more resources are added to the system. And there's also frontends with the user not currently navigated to a resource, where every 'resource' topic message is unhelpful.  
If I instead subscribe to a topic a la 'resource/1', I can have the python backend only send messages about the resource with id 1, and thus reduce the unnecessary spam of messages. But then I need to be able to unsubscribe to 'resource/1' and subscribe to e.g. 'resource/2' later, when the user has navigated to that resource.

Is there some technique of proper granularity that I'm missing here, that could avoid this unsubscribe/subscribe behaviour?


> The channel id is randomized by default, but it can also be managed by the application layer - which allows you to retain identity across connections, but keep them in account (for example for subscriptions).
It isn't used in this code; but it could be. In addition that simple RPC flow is used to verify that the RPC connection is alive and well - which is worth the effort even just for that.

To be honest, one of the reasons I didn't include the get_channel_id rpc method is that it's quite cumbersome when I'm trying to be statically typed. The channel knows about the methods, because it needs to call them. To support the get_channel_id method, the methods object also needs to reference the channel, and we have a cycle. It's not a huge problem, but bothersome when I'm trying to avoid things being potentially undefined or null.  
I'm also trying to use strict and proper typescript types, and having an object containing functions under arbitrary names/keys while also holding a reference to a channel (or channel id) isn't exactly conducive to nice static types. I'm sure I can find a solution if need be, but I was hoping to find a use case for it that was also applicable for us, or that it was entirely unnecessary.

**Reply by orweis** [reply]
> Is there some technique of proper granularity that I'm missing here, that could avoid this unsubscribe/subscribe behaviour?

I see now that your challenge here isn't a complex set of subscriptions, but dynamic subscriptions - so yeah I think updating the this dynamically is unavoidable. An option to consider is to have the app layer on top of the pub/sub manage what events are posted to each user, while each suer subscribes to their own unique topic, and then having the dynamic mapping happen there instead of in the pub/sub layer. That being said I think it's fair to use the pub/sub for this mapping.


> To support the get_channel_id method, the methods object also needs to reference the channel, and we have a cycle

How about exporting the base interface for a ChannelId the a 3rd file. 

> I'm sure I can find a solution if need be, but I was hoping to find a use case for it that was also applicable for us, or that it was entirely unnecessary.

I think it's useful, even if just for debugging - but it's not critical - as I don't think the FE implementation needs to have full feature parity with the Python backend, at least not from day one.

**Reply by Graeme22** [reply]
Hi @jhk-mjolner, how is this coming along? I'm interested in using it potentially.

**Reply by jhk-mjolner** [reply]
@Graeme22 I'm using it and it's working solidly for me.
Sadly, billable hours usually win when it comes to it vs. contributing back to OSS, so I still haven't merged anything back to permitio's codebase. I'm still open to doing it at some point, though I'm unsure where the frontend files should be placed in their repository.

But if you're interested in using my code, you should be able to simply copy from https://github.com/Open-Pectus/Open-Pectus/tree/main/frontend/src/fastapi_websocket ~~and remember to add the `unsubscribe` method on the python `PubSubEndpoint`'s `method_class` as seen here https://github.com/Open-Pectus/Open-Pectus/blob/main/openpectus/aggregator/frontend_publisher.py#L19~~ *edit: this has been included in the library.*

Having the files yourself will also allow you to extend it with features you might need, that I haven't implemented.
